File: MATLAB/RDT/StopCriteria.m
classdef StopCriteria < handle
    
    properties
        solutionNotRequired;
        maxIterations;
        maxRunTime;
    end
    
    methods
        function o = StopCriteria(solution_not_required, max_iterations, max_run_time)
            o.solutionNotRequired = solution_not_required;
            o.maxIterations = max_iterations;
            o.maxRunTime = max_run_time;
        end
        
        function do_stop = stop(o, elapsed_time, iteration_count, BSF)
            do_stop = 0;
            if (~isempty(BSF) && BSF.distToHere < Inf) || (o.solutionNotRequired)
               if (iteration_count > o.maxIterations && o.maxIterations >= 0 )...
                       || (elapsed_time > o.maxRunTime && o.maxRunTime >= 0 )
                  do_stop = 1; 
               end
            end
        end
    end
end

File: python/utils/stop_criteria.py
class StopCriteria():

	def __init__(self, solution_required, max_iterations=float('inf'), max_run_time=float('inf')):
		self.solutionRequired = solution_required
		self.maxIterations = max_iterations
		self.maxRunTime = max_run_time

	def stop(self, sol_found, iterations, time, conj = 'OR'):
		if not sol_found and self.solutionRequired:
			return False

		if conj == 'OR':
			return (iterations >= self.maxIterations) or (time>= self.maxRunTime)
		else:
			return (iterations >= self.maxIterations) and (time>= self.maxRunTime)

